### DISCUSSION TITLE
Tapping into Logger class for custom fields / Filtering based on Scenario

### DISCUSSION THREAD

**Discussion by dschibster** [jongpie/NebulaLogger#431] [Ideas]
Hey there @jongpie !

I just today received a [request](https://github.com/dschibster/sfdx-batch-orchestrator/issues/18) to integrate sfdx-batch-orchestrator with NebulaLogger, which I think is a pretty great idea but there are a few things where I would appreciate some pointers.

For the things that are currently done on Batch Apex Status Events that I send out, I should be able to simply create a Plugin given the framework you already expose, however when it comes to discoverability I am a little stumped right now so I hope you can help me a little. :) 

Logs in Batch Orchestrator are discovered by associating both the Schedule and Job Id (records in this case) to the Log Entry, grouping them based on the same Async Apex Job Id and creating the Parent before linking Children to it. It's a little finicky because the entire creation process is offset into the Platform Event.

I was trying to come up with something that can replicate this behavior somehow, and finding the Record on the FlexiPage using the Related Log Entries Component alleviates things, but it doesn't quite hit the mark. Are there plans to create a similar LWC but by e.g. using a Scenario as a parameter to input, giving out Logs? I know, asking a lot here. :D Otherwise I might just take a look for the purposes of this integration myself.

Last thing. discoverability as mentioned hinges on having similar Ids currently, which to some extent I am able to alleviate by using the Nebula Logger which creates Log + Child Entries all at once, but the biggest help would actually be to be able to tap into the LogEntryBuilder or to add additional fields to the Logger class in order to put Lookups / Custom Fields onto the Parent Log. This would help a lot in bringing everything together. Do you have any suggestions on how to tackle this issue?

Any feedback is welcome. Thanks!

**Comment by jongpie** [comment]
Hi @dschibster, thanks again for your patience on my very late response! I think there may need to be some enhancements to Nebula Logger to make your plugin easier to implement
1. Just as an FYI, if you want to have the integration with Nebula Logger be optional, you can look at how @jamessimone has built [his plugin for Apex Rollup to optionally use Nebula Logger](https://github.com/jamessimone/apex-rollup/tree/main/plugins/NebulaLogger). Essentially, you would have another unlocked package that depends on both sfdx-batch-orchestrator _and_ Nebula Logger - any orgs that want to integrate sfdx-batch-orchestrator with Nebula Logger would need to install the additional package.
2. I don't currently have any places to make the 'Related Log Entries' component (or a new component) that uses the Scenario as a parameter - I'm happy to consider adding it as a long-term enhancement, but unfortunately I don't think I will have time to do that for several months. But can you clarify on what flexipage would need to display the component? I just want to make sure that I understand the data model/relationships needed.
3. You can add custom fields within your plugin to extend Nebula Logger's data model, but I may need to make an enhancement to Nebula Logger to make your idea fully feasible. The current versions of Nebula Logger support adding custom fields to the included objects - I currently have [a Slack plugin](https://github.com/jongpie/NebulaLogger/tree/v4.9.9/nebula-logger/plugins/slack) that includes [2 new fields on the `Log__c` object](https://github.com/jongpie/NebulaLogger/tree/v4.9.9/nebula-logger/plugins/slack/plugin/core/main/log-management/objects/Log__c). But there is not currently a very clean approach for mapping custom fields between objects - and in your case, it sounds like you would need to have custom fields on `LogEntryEvent__e` (which you could populate when logging data, using `Logger.info('some log message')?.getLogEntryEvent().Some_Custom_Field__c = 'some value';`), but there's not currently a great way map the `LogEntryEvent__e` field to a `Log__c` or `LogEntry__c` field. I can add this functionality in a future release though - I started a prototype for this several months ago, but never finished implementing it. I can work on adding this once I finish a few other pending PRs.
